torch/csrc/jit/codegen/cuda/lower_shift.cpp

```diff
@@ -0,0 +1,684 @@
+  }
+}
+
+// utility function
+kir::Val* makeAddExpr(kir::Val* lhs, int rhs) {
+  if (rhs == 0) {
+    return lhs;
+  } else if (rhs > 0) {
+    kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+    return ir_builder.addExpr(lhs, ir_builder.create<kir::Int>(rhs));
+    return lhs;
+  } else {
+    kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+    return ir_builder.subExpr(lhs, ir_builder.create<kir::Int>(-rhs));
+  }
+}
+
+} // namespace
+
+void ShiftPredicateInserter::insert(
+    kir::Expr* expr,
+    const std::vector<kir::ForLoop*>& loops,
+    kir::Bool* thread_pred) {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // thread predication is not supported yet
+  TORCH_INTERNAL_ASSERT(
+      thread_pred->isConst() && thread_pred->value().value(),
+      "Thread predication is not supported for expressions with halo-extended outputs");
+
+  kir::TensorView* out_tv = nullptr;
+  for (auto out : expr->outputs()) {
+    if (out->isA<kir::TensorView>()) {
+      out_tv = out->as<kir::TensorView>();
+    }
+  }
+  TORCH_INTERNAL_ASSERT(out_tv != nullptr, "Missing kir::TensorView output");
+
+  const auto predicates = getPredicate(expr, loops, out_tv);
+  const auto shift_pred = predicates[0];
+  const auto padding_pred = predicates[1];
+
+  // If null, no specific predicate is needed.
+  if (shift_pred == nullptr) {
+    return;
+  }
+
+  // The conditional branches to create:
+  //
+  // if (shift_pred) {
+  //   consumer = producer;
+  // } else {
+  //   if (padding_pred) {
+  //     consumer = 0;
+  //   }
+  // }
+
+  auto shift_ite = ir_builder.create<kir::IfThenElse>(shift_pred);
+
+  auto& scope = loops.back()->body();
+
+  // Insert the if statement
+  scope.insert_before(expr, shift_ite);
+
+  // Remove the expr from the list
+  scope.erase(expr);
+
+  // Place the expr inside the if statement
+  shift_ite->thenBody().push_back(expr);
+
+  // Pading by zero
+  auto bounds_ite = ir_builder.create<kir::IfThenElse>(padding_pred);
+  const int pad_value = 0;
+  auto pad_expr = ir_builder.create<kir::UnaryOp>(
+      UnaryOpType::Set, out_tv, ir_builder.create<kir::Int>(pad_value));
+  bounds_ite->thenBody().push_back(pad_expr);
+  // Insert the else block
+  shift_ite->elseBody().push_back(bounds_ite);
+}
+
+std::array<kir::Bool*, 2> ShiftPredicateInserter::getPredicate(
+    const kir::Expr* expr,
+    const std::vector<kir::ForLoop*>& loops,
+    kir::TensorView* out_tv) {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  TensorView* out_fuser_tv = out_tv->fuserTv();
+
+  const bool needs_shift_predicate =
+      gpu_lower->haloInfo().needsShiftPredicate(out_fuser_tv->definition());
+
+  if (!needs_shift_predicate) {
+    return {nullptr, nullptr};
+  }
+
+  const auto& root_domain = out_fuser_tv->getRootDomain();
+
+  auto shift_expr = dynamic_cast<ShiftOp*>(out_fuser_tv->definition());
```

Comment:
Yes, it is the fusion version of expr. I haven't added `ShiftOp` in KIR but just convert it to `UnaryOp::Set`.

Reply:
Re: `->as`: 

I still like to use `dynamic_cast` for cases like this since with `as`, I'd need to write:

```
auto shift_expr = out_fuser_tv->isA<ShiftOp>() ? out_fuser_tv->as<ShiftOp>() : nullptr;
```

Reply:
Thanks, my mistake. I thought it was guaranteed to be a ShiftOp.